src/mbr.c

```diff
@@ -0,0 +1,270 @@
+
+	if (read(fd, mbr, sizeof(*mbr)) != sizeof(*mbr)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"read failed: %s", g_strerror(errno));
+		goto out;
+	}
+
+	if (mbr->magic_number[0] != MBR_MAGIC_NUMBER_L ||
+	    mbr->magic_number[1] != MBR_MAGIC_NUMBER_H) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"no valid master boot record found");
+		goto out;
+	}
+
+	res = TRUE;
+out:
+	return res;
+}
+
+static gboolean mbr_is_part_overlapping(guint32 start, guint32 size,
+		struct partition_tbl_entry *partition_tbl, GError **error)
+{
+	guint32 entry_end;
+	guint i;
+
+	for (i = 0; i < MBR_NUMBER_OF_PARTITIONS; i++) {
+		entry_end = partition_tbl[i].partition_start +
+		            partition_tbl[i].partition_size - 1;
+
+		if (start >= partition_tbl[i].partition_start && start <= entry_end)
+			break;
+
+		if (partition_tbl[i].partition_start >= start &&
+		    partition_tbl[i].partition_start <= start + size - 1)
+			break;
+	}
+
+	if (i < MBR_NUMBER_OF_PARTITIONS) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"inactive boot partition collides with "
+				"partition[%d]", i);
+		return TRUE;
+	}
+
+	return FALSE;
+}
+
+gboolean r_mbr_get_non_active_boot_partition(const gchar *device,
+		struct mbr_partition *partition,
+		GError **error)
+{
+	gboolean res = FALSE;
+	struct mbr mbr;
+	GError *ierror = NULL;
+	guint32 start = 0, size;
+	guint sector_size;
+	struct partition_tbl_entry *boot_part;
+	gint fd;
+
+	fd = g_open(device, O_RDONLY);
+
+	res = r_mbr_read_mbr(fd, &mbr, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(error, ierror, "failed to read mbr:");
+		goto out;
+	}
+
+	res = FALSE;
+
+	boot_part = &mbr.partition_table[BOOT_PARTITION_ENTRY];
+
+	if (boot_part->partition_start == 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"no boot partition found in entry %d",
+				BOOT_PARTITION_ENTRY);
+		goto out;
+	}
+
+
+	if (boot_part->partition_start == boot_part->partition_size) {
+		start = 2 * boot_part->partition_size;
+	}
+	else if (boot_part->partition_start == 2 * boot_part->partition_size) {
+		start = boot_part->partition_size;
+	}
+	else {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"boot partition's start address does not equal "
+				"it's size or twice it's size");
+		goto out;
+	}
+
+	size = boot_part->partition_size;
+
+	/* check if new boot partition overlaps with any other partition */
+	if(mbr_is_part_overlapping(start, size, mbr.partition_table, &ierror)) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	if(!get_sectorsize(fd, &sector_size))
+		sector_size = DEFAULT_SECTOR_SIZE;
+
+	partition->start = (guint64)start * sector_size;
+	partition->size = (guint64)size * sector_size;
+
+	res = TRUE;
+out:
+	if (fd >= 0)
+		g_close(fd, NULL);
+
+	return res;
+}
+
+static void r_mbr_get_chs(guint8 *chs, guint32 sector,
+		const struct hd_geometry *geometry)
+{
+	chs[2] = sector % geometry->sectors + 1;
```

Comment:
Is the "offical" way to do this documented somewhere? I'd like a comment here, describing how you got this calculation.